### ISSUE TITLE
Design Proposal: Federated Job: Cross-cluster task rebalancing

### ISSUE THREAD

**Issue by quinton-hoole** [kubernetes/kubernetes#45187]
Figure out how best to do task rebalancing across clusters for Federated Jobs.

See #36197 for background.

@jianhuiz leads this effort.  Assigning myself as proxy, as he is not part of the K8s org yet. 

@kubernetes/sig-federation-feature-requests

**Comment by quinton-hoole** [comment]
From @jianhuiz 

1. how to decide a re-balancing is needed (succeeded ==0 and failed >0? failed = 0 and active ==0 for a while?; active is far less than parallelism?; jobs on a cluster can succeed but significantly slower than other clusters)
2. how to make an efficient re-balance. should the succeeded count being considered when re-planning?

**Comment by jianhuiz** [comment]
From: https://docs.google.com/document/d/1TZ4ZMZ8AEOlVyK6H_q-zNpAdlzTip8sxMQUVQeQ-O-w/edit#

@David Oppenheimer
My understanding is that what's wanted here is "considered all nodes and it
can't schedule onto any of them"? We do surface that in PodStatus today in
PodReasonUnschedulable. In terms of how the federated job controller can
make use of this information without having to track each pod, I think the
missing piece is https://github.com/kubernetes/kubernetes/issues/32863.
Once that's implemented, this kind of information should be surfaced in the
JobStatus, so the federated job controller only needs to watch the
per-cluster Jobs, which IIUC it is doing anyway. Until
https://github.com/kubernetes/kubernetes/issues/32863 is implemented, I
think you would just have to periodically fetch the list of pending pods
from the underlying cluster, extract the ones that are unschedulable (by
examining PodReasonUnschedulable), find which of them were created by a
local instance of a federated Job, and then take action based on that.

@quinton-hoole 
I think a possibly sensible approach here might be to start by creating Jobs in the underlying clusters with only a single worker, and incrementally scaling them up (possibly one worker at a time) only once it has been demonstrated that workers are able to schedule in that cluster. That way, workers will be allocated to the clusters where they are actually able to schedule. Cluster-level Jobs that end up with a single pending pod will remain in that state (approximately) until either additional capacity becomes available, or the required workers are successfully scheduled into other clusters.

**Comment by irfanurrehman** [comment]
This issue was labelled only for sig/multicluster and is thus moved over to kubernetes/federation#133.
If this does not seem to be right, please reopen this and notify us @kubernetes/sig-multicluster-misc.
/close
